doc: adds mempool v1 protocol spec (#1108)
celestiaorg/celestia-core: spec/p2p/reactors/mempool-v1.md

> ## Mempool Reactor Overview

Author: Following your comment, I ran markdown lint on the doc and fixed all the problems, please see https://github.com/celestiaorg/celestia-core/pull/1108/commits/e26723e7019b886c59a206dc7cfbe7b232857a61.

> The Celestia-core's p2p layer, which is a fork of CometBFT, consists of channels and reactors. Peers establish connections within specific channels, effectively forming peer-to-peer groups (each channel represents such a group). The transmission of messages within a channel is governed by the associated reactor, essentially containing the protocol rules for that channel.

Author: > Reactors connect to one or more channels. Reactors define the network traffic rules for a channel.

There is only one reactor per channel (think of channel as the name of the protocol, and the reactor as the logic of that protocol). 

> Peers join channels and connect to all (?) other peers that have joined that channel.

That is right, however, a peer does not connect to all the other peers in that channel, rather connects to a subset of them. 

Adding a bit more context: When a new peer connects to another peer that is already part of the Celestia-core network, they exchange their supported channel IDs and initiate message exchange for the channels they both support. 

Please let me know if that resolves your confusion.

One notable channel is the mempool channel, identified as [`MempoolChannel`](https://github.com/celestiaorg/celestia-core/blob/3f3b7cc57f5cfc5e846ce781a9a407920e54fb72/mempool/mempool.go#L14) with a specific channel ID of [`0x30`](https://github.com/celestiaorg/celestia-core/blob/3f3b7cc57f5cfc5e846ce781a9a407920e54fb72/mempool/mempool.go#L14). The mempool reactor manages the dissemination of transactions across the network. It's important to highlight that there's a direct correspondence between reactors and the channels they are connected to. Consequently, the mempool reactor is the exclusive reactor linked to the mempool channel with ID `0x30`. This document will provide an overview of the protocol implemented by the mempool v1 reactor along with its traffic analysis.

## Transaction Flow and Lifecycle in a Node
A node can receive a transaction through one of two pathways: either a user initiates the transaction directly to the node, or the node acquires a transaction from another peer. Upon receiving a transaction, the following steps occur:

> 1. The transaction's validity is assessed, and if it passes the validation criteria, it is added to the mempool. Furthermore, the transaction's height is set to match the current block height.

Author: I added further clarification in the doc, please see https://github.com/celestiaorg/celestia-core/pull/1108/commits/4b678a1a0ebff92e7b3e31b87efd0f772a63bca3. 
Please let me know if more explanation is needed.

2. **Peer Tracking**: In the event that the transaction originates from another peer, the sending peer is marked to prevent redundant transmission of the same transaction.

>    Subsequently, there are two concurrent processes underway:

Author: Clarified in this commit https://github.com/celestiaorg/celestia-core/pull/1108/commits/12ef125e14ad3c8172fef8ca46058077c9f63ad0.

    - When a [block is committed](https://github.com/celestiaorg/celestia-core/blob/367caa33ef5ab618ea357189e88044dbdbd17776/state/execution.go#L324):
        - the transactions within that block that are successfully delivered to the app are removed from the mempool ([ref](https://github.com/celestiaorg/celestia-core/blob/993c1228977f206c80cb0f87ac1d4f002826e904/mempool/v1/mempool.go#L418)). They are also placed in the mempool cache ([ref](https://github.com/celestiaorg/celestia-core/blob/993c1228977f206c80cb0f87ac1d4f002826e904/mempool/v1/mempool.go#L411-L412)).
        - The remaining transactions are subjected to two checks:
            - their Time-to-Live (TTL) is examined ([ref](https://github.com/celestiaorg/celestia-core/blob/993c1228977f206c80cb0f87ac1d4f002826e904/mempool/v1/mempool.go#L421)), and any transactions that have expired are promptly removed from the mempool ([ref](https://github.com/celestiaorg/celestia-core/blob/993c1228977f206c80cb0f87ac1d4f002826e904/mempool/v1/mempool.go#L743)).

>             - Next, the remaining transactions are re-evaluated for validity against the updated state ([ref](https://github.com/celestiaorg/celestia-core/blob/993c1228977f206c80cb0f87ac1d4f002826e904/mempool/v1/mempool.go#L429-L430)) duo to the mempool [`recheck` config](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L708) that is set to `true` ([ref](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L761)). Any transactions that are found to be invalid are removed from the mempool.

Author: Fixed https://github.com/celestiaorg/celestia-core/pull/1108/commits/745e390913db0855477ee04eeb35692c91ffc120

    - A copy of the transaction is dispatched to that peer if the peer
        -  is online
        - supports the mempool channel ID ([ref](https://github.com/celestiaorg/celestia-core/blob/ad660fee8f186d6f7e5e567ea23ea813f5038d90/p2p/peer.go#L319))
        - has a height difference of one (meaning it lags behind the transaction by a single block). If the height difference is greater, a waiting period is observed to allow the peer to catch up ([ref](https://github.com/celestiaorg/celestia-core/blob/64cd9ab7c67c945d755fb4fbd5afb2d352874eea/mempool/v1/reactor.go#L286-L289)).
    - **Peer Tracking**: Each transaction is sent to a peer only once, and the recipient peer is marked to prevent the retransmission of the same transaction ([ref](https://github.com/celestiaorg/celestia-core/blob/64cd9ab7c67c945d755fb4fbd5afb2d352874eea/mempool/v1/reactor.go#L304)).

## Constraints and  Configurations
The relevant constraints and configurations for the mempool are as follows ([ref](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L758)):

> - `Size`: This parameter specifies the total number of transactions that the mempool can hold, with a maximum value of `5000`.

Author: Yes, it is a configurable parameter (the link is also provided in the intro line), the core uses `5000` as the default size. It can be adjusted in the config.toml file.

> - `Size`: This parameter specifies the total number of transactions that the mempool can hold, with a maximum value of `5000`.

Author: I have expanded on this aspect in the following commit https://github.com/celestiaorg/celestia-core/pull/1108/commits/5d7194c92daae3bae36dc57772e5c0ba8345dc65.

> -  `MaxTxsBytes`: The `MaxTxsBytes` parameter defines the maximum size of the mempool in bytes, with a limit of `1GB`  by default.

Reviewer: [nit] "with a limit of `1GB` by default" confuses me. Is 1GB a hard-coded upper bound or the default value of the parameter. Proposal to rephrase to:

```suggestion
-  `MaxTxsBytes`: The `MaxTxsBytes` parameter defines the maximum size of the mempool in bytes, with a default value of `1GB`.
```

Author: Sure, addressed https://github.com/celestiaorg/celestia-core/pull/1108/commits/a3b9d1ae5192f52cbe48f26fd4baaa9bc0fadf2b.

- `MaxTxBytes`: The `MaxTxBytes` parameter specifies the maximum size of an individual transaction, which is set to `1MB`.
-  `TTLNumBlocks` and `TTLDuration` : These settings determine the number of blocks and time after which a transaction is removed from the mempool if it has not been included in a block. The default is set to zero, however, on [celestia-app side](https://github.com/celestiaorg/celestia-app/blob/ccfb3e5e87d05d75a92ad85ab199d4f0c4879a0a/app/default_overrides.go#L221-L222) these values are over-written to `5` and `5*15 s`, respectively.

For each peer-to-peer connection, the following limits apply to the aggregate traffic rate of all the channels ([ref](https://github.com/celestiaorg/celestia-core/blob/3f3b7cc57f5cfc5e846ce781a9a407920e54fb72/libs/flowrate/flowrate.go#L177)):

> -  `SendRate`: The `SendRate` parameter enforces a default average sending rate of [`5120000 B= 5MB/s`](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L615). It ensures that data is sent at this maximum rate. This parameter does not seem to be overwritten by the celestia-app.

Reviewer: [question] how can this rate simultaneously be an average and a maximum rate?

Author: I have an inline comment, in later lines about clarifying whether these rates are the max or the average. I am going to remove the term 'average' temporarily until I can confirm whether the rates refer to the average rate or the maximum rate. Please see https://github.com/celestiaorg/celestia-core/pull/1108/commits/a047f679831ddc7664b9c58a342ee5d7ae2c8de2.

> - `RecvRate`: The `RecvRate` parameter enforces a default average receiving rate of [`5120000 B= 5MB/s`](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L616). It ensures that data is received at this maximum rate. This parameter does not seem to be overwritten by the celestia-app.

Reviewer: [question] how can this rate simultaneously be an average and a maximum rate?

Author: I have covered it here https://github.com/celestiaorg/celestia-core/pull/1108#discussion_r1346573263.

Other P2P configs ([ref](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L524)) that would be relevant to the traffic analysis are:
- [`max_num_inbound_peers` and `max_num_outbound_peers`](https://github.com/celestiaorg/celestia-core/blob/37f950717381e8d8f6393437624652693e4775b8/config/config.go#L604-L605): These parameters indicate the total number of inbound and outbound peers, respectively. The default values are `40` for inbound peers and `10` for outbound peers ([excluding persistent peers](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L553-L554)).
<!-- The allocation of max_num_inbound_peers and max_num_outbound_peers across various connection types requires clarification. For instance, whether max_num_inbound_peers includes both unconditional peers and persistent peers or not. -->

## Traffic Rate Analysis
In the analysis provided below, we consider the knowledge of the following network parameters
- `d`: Node degree (total incoming and outgoing connections)

> <!-- - transaction rate: `transaction_rate` total number of transactions per second submitted to the network -->

Author: Yes, done https://github.com/celestiaorg/celestia-core/pull/1108/commits/636f4587d97c730fc135e8c0c299cfd83dc67e6f.

- `transaction_rate` which specifies that total size of transactions in bytes per second submitted to the network.

> - `C`: total number of connections in the network

Author: I was thinking of extending the analysis to encompass the entire network, in which case  `C` would be a relevant parameter. However, I can omit it for now as I haven't covered that section yet.

> - `C`: total number of connections in the network

Author: Fixed https://github.com/celestiaorg/celestia-core/pull/1108/commits/3241e179d2b6f7402f08622c8c540cdff7f6a04f.

Transactions are assumed to comply with the transaction size, are valid and are accepted by the mempool. We also assume all the peers are up and running.

### Traffic Rate Analysis for a Node
We distinguish between the incoming and outgoing traffic rate, and denote them by  `incoming_traffic_rate` and  `outgoing_traffic_rate`, respectively.
- **Worst case scenario**: a transaction is exchanged by the two ends of
connection simultaneously, contributing to both incoming and outgoing traffic.
In a network, with transaction rate `transaction_rate` and a node with `d` degree, the traffic rates are calculated as follows:
`incoming_traffic_rate = d * transaction_rate`
`outgoing_traffic_rate = d * transaction_rate`

These max rates are further constrained by the `SendRate` and `RecRate`.
`incoming_traffic_rate = d * min(transaction_rate, SendRate)`

> `outgoing_traffic_rate = d * min(transaction_rate, RecRate)`

Author: Indeed, it is a typo, fixed it now https://github.com/celestiaorg/celestia-core/pull/1108/commits/ff9dc269b9dd74a28920470774e011fe6c7db71d.

- **Best case scenario**: a transaction is exchanged only once, contributing to either incoming or outgoing traffic. This is because both ends of the connection keep track of the transactions they have seen on a connection (whether via sending or receiving). If one peer sends a transaction before the other, they both mark it as sent/received, ensuring they do not redundantly transmit it to each other.
In a network, with transaction rate `transaction_rate` and a node with  degree of `d`, the node's traffic rate in best case would be:
`traffic_rate (=incoming_traffic_rate + outgoing_traffic_rate) = d * transaction_rate`

We can draw the following conclusions (to be extended and verified):
- With a known given transaction rate `transaction_rate`, a node's (in + out) traffic rate should range from `d * transaction_rate` to `2 * d * transaction_rate`.

> - To handle a particular `transaction_rate` (network throughput), the node's `SendRate` and `RecRate` should be at least `transaction_rate` to handle the worst case scenario (this is only to undertake the load incurred by the mempool reactor).

Reviewer: [nit] here `RecRate` is used but it appears defined as `RecvRate` above. Proposal for it to be consistent.

Author: Fixed https://github.com/celestiaorg/celestia-core/pull/1108/commits/ff9dc269b9dd74a28920470774e011fe6c7db71d.

4. **Broadcast Process**:
   For each peer and for every transaction residing in the mempool, the following actions are taken ([ref](https://github.com/celestiaorg/celestia-core/blob/64cd9ab7c67c945d755fb4fbd5afb2d352874eea/mempool/v1/reactor.go#L244)):
    - A copy of the transaction is dispatched to that peer if the peer
      - is online
      - supports the mempool channel ID ([ref](https://github.com/celestiaorg/celestia-core/blob/ad660fee8f186d6f7e5e567ea23ea813f5038d90/p2p/peer.go#L319))
      - has a height difference of one (meaning it lags behind the transaction by a single block). If the height difference is greater, a waiting period is observed to allow the peer to catch up ([ref](https://github.com/celestiaorg/celestia-core/blob/64cd9ab7c67c945d755fb4fbd5afb2d352874eea/mempool/v1/reactor.go#L286-L289)).
    - **Peer Tracking**: Each transaction is sent to a peer only once, and the recipient peer is marked to prevent the retransmission of the same transaction ([ref](https://github.com/celestiaorg/celestia-core/blob/64cd9ab7c67c945d755fb4fbd5afb2d352874eea/mempool/v1/reactor.go#L304)).

> # Constraints and  Configurations

Reviewer: ```suggestion
## Constraints and  Configurations
```

<!-- Depending on the state of this [PR](https://github.com/celestiaorg/celestia-app/pull/2390) we may have further constraints on the bandwidth. -->

Other P2P configs ([ref](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L524)) that would be relevant to the traffic analysis are:

- [`max_num_inbound_peers` and `max_num_outbound_peers`](https://github.com/celestiaorg/celestia-core/blob/37f950717381e8d8f6393437624652693e4775b8/config/config.go#L604-L605): These parameters indicate the total number of inbound and outbound peers, respectively. The default values are `40` for inbound peers and `10` for outbound peers ([excluding persistent peers](https://github.com/celestiaorg/celestia-core/blob/2f93fc823f17c36c7090f84694880c85d3244764/config/config.go#L553-L554)).
-
<!-- The allocation of max_num_inbound_peers and max_num_outbound_peers across various connection types requires clarification. For instance, whether max_num_inbound_peers includes both unconditional peers and persistent peers or not. -->

> # Traffic Rate Analysis

Reviewer: ```suggestion
## Traffic Rate Analysis
```
